### PR TITLE
8311596: Add separate system properties for TLS server and client for maximum chain length

### PR DESCRIPTION
**Pull request by haimaychao**
Please review the enhancement for JDK-8311596 and its CSR JDK-8313236. Thank you.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change requires CSR request [JDK-8313236](https://bugs.openjdk.org/browse/JDK-8313236) to be approved

### Issues
 * [JDK-8311596](https://bugs.openjdk.org/browse/JDK-8311596): Add separate system properties for TLS server and client for maximum chain length (**Enhancement** - P3)
 * [JDK-8313236](https://bugs.openjdk.org/browse/JDK-8313236): Add separate system properties for TLS server and client for maximum chain length (**CSR**)

### Reviewers
 * [Jamil Nimeh](https://openjdk.org/census#jnimeh) (@jnimeh - **Reviewer**) ⚠️ Review applies to [e8174347](https://git.openjdk.org/jdk/pull/15163/files/e8174347dc53c0f6e4eaf78680f230a9a5d12467)
 * [Weijun Wang](https://openjdk.org/census#weijun) (@wangweij - **Reviewer**) ⚠️ Review applies to [b3192df3](https://git.openjdk.org/jdk/pull/15163/files/b3192df38cb13286b573dfe4101553b1a68d4dfc)
 * [Sean Mullan](https://openjdk.org/census#mullan) (@seanjmullan - **Reviewer**) ⚠️ Review applies to [b3192df3](https://git.openjdk.org/jdk/pull/15163/files/b3192df38cb13286b573dfe4101553b1a68d4dfc)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/15163/head:pull/15163` \
`$ git checkout pull/15163`

Update a local copy of the PR: \
`$ git checkout pull/15163` \
`$ git pull https://git.openjdk.org/jdk.git pull/15163/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 15163`

View PR using the GUI difftool: \
`$ git pr show -t 15163`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/15163.diff">https://git.openjdk.org/jdk/pull/15163.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/15163#issuecomment-1665964301)

### CODE REVIEW

File: src/java.base/share/classes/sun/security/ssl/SSLConfiguration.java [hunk 1]
```diff
@@ -112,6 +112,12 @@ final class SSLConfiguration implements Cloneable {
     static final int maxCertificateChainLength = GetIntegerAction.privilegedGetProperty(
             "jdk.tls.maxCertificateChainLength", 10);
 
+    // Limit the maximum certificate chain length accepted from clients
```
Comment (mcpowers) [inline src/java.base/share/classes/sun/security/ssl/SSLConfiguration.java:115]: Should these be moved to after line 89?

File: src/java.base/share/classes/sun/security/ssl/SSLConfiguration.java [hunk 2]
```diff
@@ -133,6 +139,25 @@ final class SSLConfiguration implements Cloneable {
...
 
+    /*
+     * If the jdk.tls.maxClientCertificateChainLength or jdk.tls.maxServerCertificateChainLength
+     * system property is not set, it will default to the value of the
+     * jdk.tls.maxCertificateChainLength system property. If either of them is
+     * set, it will supersede the value of the jdk.tls.maxCertificateChainLength
+     * system property.
+     */
+    static {
+        Integer clientLen = GetIntegerAction.privilegedGetProperty(
```
Comment (mcpowers) [inline src/java.base/share/classes/sun/security/ssl/SSLConfiguration.java:150]: I think you could call `privilegedGetProperty` with the default value as second argument.

**Review by mcpowers — COMMENTED** [review]
All of the other changes look fine to me.
Does this need a test?

File: src/java.base/share/classes/sun/security/ssl/SSLConfiguration.java [hunk 3]
```diff
@@ -133,6 +139,25 @@ final class SSLConfiguration implements Cloneable {
...
+                "jdk.tls.maxClientCertificateChainLength");
+        maxClientCertificateChainLength = (clientLen != null) ?
+                clientLen : maxCertificateChainLength;
+
+        Integer serverLen = GetIntegerAction.privilegedGetProperty(
+                "jdk.tls.maxServerCertificateChainLength");
+        maxServerCertificateChainLength = (serverLen != null) ?
+                serverLen : maxCertificateChainLength;
+    }
```
Comment (jnimeh) [inline src/java.base/share/classes/sun/security/ssl/SSLConfiguration.java:159]: I wonder if we should take the opportunity here with these new properties as well as `jdk.tls.maxCertificateChainLength` to also equate negative numbers (and maybe zero) to be the default.  Right now only property values that fail the internal parseInt conversion will evaluate to `null` and would be assigned the default I think.  But a negative value I think would be taken as-is from the property.  Should a negative max cert chain length get set to the default?  If so, it might also make sense to give a warning about the offending value and note that it is being set to the default (similar to what `GetPropertyAction.privilegedGetTimeoutProp()` does).
If you think this is worthwhile, the CSR should probably be updated to reflect that also.

File: src/java.base/share/classes/sun/security/ssl/SSLConfiguration.java (hunk 1 above)
Comment (haimaychao) [reply src/java.base/share/classes/sun/security/ssl/SSLConfiguration.java:115]: Remain to stay after the related maxCertificateChainLength.

File: src/java.base/share/classes/sun/security/ssl/SSLConfiguration.java (hunk 2 above)
Comment (haimaychao) [reply src/java.base/share/classes/sun/security/ssl/SSLConfiguration.java:150]: Done.

File: src/java.base/share/classes/sun/security/ssl/SSLConfiguration.java (hunk 3 above)
Comment (haimaychao) [reply src/java.base/share/classes/sun/security/ssl/SSLConfiguration.java:159]: Change made to set to the default when a negative value is set for these system properties. Updated CSR for this.

**Review by jnimeh — APPROVED** [review]
This looks good to me.  Thanks for making that negative value update.

File: src/java.base/share/classes/sun/security/ssl/SSLConfiguration.java [hunk 4]
```diff
@@ -108,9 +108,14 @@ final class SSLConfiguration implements Cloneable {
...
-    // Set the max certificate chain length to 10
-    static final int maxCertificateChainLength = GetIntegerAction.privilegedGetProperty(
-            "jdk.tls.maxCertificateChainLength", 10);
+    // Maximum allowed certificate chain length
+    static final int maxCertificateChainLength;
+
+    // Limit the maximum certificate chain length accepted from clients
+    static final int maxClientCertificateChainLength;
+
+    // Limit the maximum certificate chain length accepted from servers
```
Comment (mcpowers) [inline src/java.base/share/classes/sun/security/ssl/SSLConfiguration.java:117]: I would drop "maximum" and just say "Limit the certificate chain length accepted from servers". It's not worth making this change unless you have to make another change to this file.

Comment (haimaychao) [reply src/java.base/share/classes/sun/security/ssl/SSLConfiguration.java:117]: ok

File: src/java.base/share/classes/sun/security/ssl/SSLConfiguration.java [hunk 5]
```diff
@@ -133,6 +138,50 @@ final class SSLConfiguration implements Cloneable {
...
+        /*
+         * When jdk.tls.maxCertificateChainLength is set, it will override
+         * the defaults for jdk.tls.maxClientCertificateChainLength and
+         * jdk.tls.maxServerCertificateChainLength if they are not explicitly
+         * overridden. That means if either of those properties have been set,
+         * the jdk.tls.maxCertificateChainLength property will not override
+         * the values.
+         */
+        if (maxCertificateChainLength > 0) {
+            if (clientLen == 8) {
```
[the remark this answers is not in the document]
Comment (haimaychao) [reply src/java.base/share/classes/sun/security/ssl/SSLConfiguration.java:173]: Since 8 is the default for "jdk.tls.maxClientCertificateChainLength", it is going to be overridden when "jdk.tls.maxCertificateChainLength" is set. Setting "jdk.tls.maxClientCertificateChainLength" to 8 is treated as keeping the original default like no-op.

[the remark this answers is not in the document]
Comment (haimaychao) [reply src/java.base/share/classes/sun/security/ssl/SSLConfiguration.java:173]: `jdk.tls.maxCertificateChainLength` will only override `jdk.tls.maxClientCertificateChainLength` if `jdk.tls.maxCertificateChainLength` is set AND `jdk.tls.maxClientCertificateChainLength` is using the default. For the case your provided here, `jdk.tls.maxClientCertificateChainLength` will be overridden to be 4 which is set by `jdk.tls.maxCertificateChainLength`.

[the remark this answers is not in the document]
Comment (seanjmullan) [reply src/java.base/share/classes/sun/security/ssl/SSLConfiguration.java:173]: Yes, I agree that if the application sets jdk.tls.maxClientCertificateChainLength or jdk.tls.maxServerCertificateChainLength, it should always take precedence even if the specified value is the same as the default. This would also preserve application behavior if one day we want to change the default to something else.

You will need to first see if these properties are set before assigning the default value.

[the remark this answers is not in the document]
Comment (haimaychao) [reply src/java.base/share/classes/sun/security/ssl/SSLConfiguration.java]: Fixed.

**Comment by XueleiFan** [comment]
I was wondering, if it is easier to learn and remember/search by following the naming style "jdk.tls.client.XXX" or "jdk.tls.server.XXX" in SunJSSE provider?

File: src/java.base/share/classes/sun/security/ssl/SSLConfiguration.java (hunk 2 above)
Comment (haimaychao) [reply src/java.base/share/classes/sun/security/ssl/SSLConfiguration.java:157]: I did update the code with his suggestion, and please see it in webrev **01: Incremental.**  However, I further change the code based on the updated CSR, that the new client and server properties have different default values now  from the existing property, and the need to know whether the new property is explicitly set or not. We want to make sure the new properties are not overridden when they are set.

**Comment by haimaychao** [comment]
@XueleiFan The current properties named` jdk.tls.client.*` and `jdk.tls.server.*` apply to settings either on the client or the server, so we'd have to rename the properties here. My suggestion is to:

- Change `jdk.tls.maxServerCertificateChainLength` to `jdk.tls.client.maxAcceptedCertificateChainLength`

- Change `jdk.tls.maxClientCertificateChainLength` to `jdk.tls.server.maxAcceptedCertificateChainLength`

Thanks!

**Comment by XueleiFan** [comment]
> @XueleiFan The current properties named` jdk.tls.client.*` and `jdk.tls.server.*` apply to settings either on the client or the server, so we'd have to rename the properties here. My suggestion is to:
> 
> * Change `jdk.tls.maxServerCertificateChainLength` to `jdk.tls.client.maxAcceptedCertificateChainLength`
> * Change `jdk.tls.maxClientCertificateChainLength` to `jdk.tls.server.maxAcceptedCertificateChainLength`
> 
> Thanks!

For the name "jdk.tls.maxServerCertificateChainLength", it is not clear to me which side, client or server, the property should be applied to.   It could also means that server can only send out certification with this limitation.

For the name `jdk.tls.client.maxAcceptedCertificateChainLength`, it could be confused to parse the word "accepted".  It could means that the accepted cert-chain length for sending out certificates.

Maybe, you can have a try with "jdk.tls.client.maxServerCertificateChainLength", which mean for client side, the server certificate chain length is limited.

**Comment by haimaychao** [comment]
It sounds good to me that the word "Accepted" is replaced with "Inbound". It should clear out the confusion I think. Thanks for the suggestion.

So how about changing the properties names from:
`jdk.tls.maxServerCertificateChainLength`  --> `jdk.tls.client.maxInboundCertificateChainLength`
`jdk.tls.maxClientCertificateChainLength` --> `jdk.tls.server.maxInboundCertificateChainLength`

File: src/java.base/share/classes/sun/security/ssl/SSLConfiguration.java [hunk 6]
```diff
@@ -108,9 +108,14 @@ final class SSLConfiguration implements Cloneable {
     static final int maxHandshakeMessageSize = GetIntegerAction.privilegedGetProperty(
             "jdk.tls.maxHandshakeMessageSize", 32768);
 
-    // Set the max certificate chain length to 10
-    static final int maxCertificateChainLength = GetIntegerAction.privilegedGetProperty(
-            "jdk.tls.maxCertificateChainLength", 10);
+    // Maximum allowed certificate chain length
+    static final int maxCertificateChainLength;
```
Comment (seanjmullan) [inline src/java.base/share/classes/sun/security/ssl/SSLConfiguration.java:112]: I think this can just be a local variable now inside the static block.

Comment (haimaychao) [reply src/java.base/share/classes/sun/security/ssl/SSLConfiguration.java]: Good catch. Done.

File: src/java.base/share/classes/sun/security/ssl/SSLConfiguration.java [hunk 7]
```diff
@@ -133,6 +138,60 @@ final class SSLConfiguration implements Cloneable {
...
+        } else {
+            globalPropSet = true;
+        }
+        maxCertificateChainLength = certLen;
+
+        // Default for jdk.tls.server.maxInboundCertificateChainLength is 8
+        Integer inboundClientLen = GetIntegerAction.privilegedGetProperty(
+                "jdk.tls.server.maxInboundCertificateChainLength");
+        if (inboundClientLen == null || inboundClientLen < 0) {
+            inboundClientLen = 8;
```
[the remark this answers is not in the document]
Comment (seanjmullan) [reply src/java.base/share/classes/sun/security/ssl/SSLConfiguration.java:157]: Hmm, but how does this work? The inbound properties override the global property if both are set.

File: src/java.base/share/classes/sun/security/ssl/SSLConfiguration.java [hunk 8]
```diff
@@ -133,6 +135,60 @@ final class SSLConfiguration implements Cloneable {
...
+        }
+
+        /*
+         * When jdk.tls.maxCertificateChainLength is set, it will override
+         * the defaults for jdk.tls.server.maxInboundCertificateChainLength and
+         * jdk.tls.client.maxInboundCertificateChainLength if they are not explicitly
+         * set. That means if either of those properties have been set,
+         * the jdk.tls.maxCertificateChainLength property will not override
+         * the values.
+         */
```
[the remark this answers is not in the document]
Comment (seanjmullan) [reply src/java.base/share/classes/sun/security/ssl/SSLConfiguration.java:178]: I agree that wording is more clear. We should also update the RN with that wording.

File: src/java.base/share/classes/sun/security/ssl/SSLConfiguration.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (haimaychao) [reply src/java.base/share/classes/sun/security/ssl/SSLConfiguration.java:157]: Updated as suggested. Changed code to:
```
        if (inboundServerLen == null || inboundServerLen < 0) {
            maxInboundServerCertChainLen = globalPropSet ? maxCertificateChainLength : 10;
        } else {
            maxInboundServerCertChainLen = inboundServerLen;
        }
```

File: src/java.base/share/classes/sun/security/ssl/SSLConfiguration.java [hunk 9]
```diff
@@ -133,6 +138,60 @@ final class SSLConfiguration implements Cloneable {
...
+
+        // jdk.tls.maxCertificateChainLength property has no default
+        Integer certLen = GetIntegerAction.privilegedGetProperty(
+                "jdk.tls.maxCertificateChainLength");
+        if (certLen == null || certLen < 0) {
+            certLen = 0;
+        } else {
+            globalPropSet = true;
+        }
+        maxCertificateChainLength = certLen;
```
[the remark this answers is not in the document]
Comment (haimaychao) [reply src/java.base/share/classes/sun/security/ssl/SSLConfiguration.java]: Removed them.

File: src/java.base/share/classes/sun/security/ssl/SSLConfiguration.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (haimaychao) [reply src/java.base/share/classes/sun/security/ssl/SSLConfiguration.java:178]: This section of comments was taken from the CSR. I updated the comments as follows. If it looks fine, I will update the related doc. Thanks!
```
        /*
         * If either jdk.tls.server.maxInboundCertificateChainLength or
         * jdk.tls.client.maxInboundCertificateChainLength is set, it will
         * override jdk.tls.maxCertificateChainLength, regardless of whether
         * jdk.tls.maxCertificateChainLength is set or not.
         * If neither jdk.tls.server.maxInboundCertificateChainLength nor
         * jdk.tls.client.maxInboundCertificateChainLength is set, the behavior
         * depends on the setting of jdk.tls.maxCertificateChainLength. If
         * jdk.tls.maxCertificateChainLength is set, it falls back to that
         * value; otherwise, it defaults to 8 for
         * jdk.tls.server.maxInboundCertificateChainLength
         * and 10 for jdk.tls.client.maxInboundCertificateChainLength.
         * Usesrs can independently set either
         * jdk.tls.server.maxInboundCertificateChainLength or
         * jdk.tls.client.maxInboundCertificateChainLength.
         */ 
```

File: src/java.base/share/classes/sun/security/ssl/SSLConfiguration.java [hunk 10]
```diff
@@ -137,56 +137,49 @@ final class SSLConfiguration implements Cloneable {
...
+         * override jdk.tls.maxCertificateChainLength, regardless of whether
+         * jdk.tls.maxCertificateChainLength is set or not.
+         * If neither jdk.tls.server.maxInboundCertificateChainLength nor
+         * jdk.tls.client.maxInboundCertificateChainLength is set, the behavior
+         * depends on the setting of jdk.tls.maxCertificateChainLength. If
+         * jdk.tls.maxCertificateChainLength is set, it falls back to that
+         * value; otherwise, it defaults to 8 for
+         * jdk.tls.server.maxInboundCertificateChainLength
+         * and 10 for jdk.tls.client.maxInboundCertificateChainLength.
+         * Usesrs can independently set either
```
Comment (seanjmullan) [inline src/java.base/share/classes/sun/security/ssl/SSLConfiguration.java:160]: Typo: s/Usesrs/Users

**Review by seanjmullan — APPROVED** [review]

File: src/java.base/share/classes/sun/security/ssl/SSLConfiguration.java (hunk 10 above)
Comment (haimaychao) [reply src/java.base/share/classes/sun/security/ssl/SSLConfiguration.java:160]: Fixed

**Comment by haimaychao** [comment]
Thanks all for the review.

File: src/java.base/share/classes/sun/security/ssl/SSLConfiguration.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (XueleiFan) [reply src/java.base/share/classes/sun/security/ssl/SSLConfiguration.java]: Sorry, I did not get time to review this behavior update.

> This section of comments was taken from the CSR. I updated the comments as follows. If it looks fine, I will update the related doc. Thanks!
> 
> ```
>         /*
>          * If either jdk.tls.server.maxInboundCertificateChainLength or
>          * jdk.tls.client.maxInboundCertificateChainLength is set, it will
>          * override jdk.tls.maxCertificateChainLength, regardless of whether
>          * jdk.tls.maxCertificateChainLength is set or not.
I'm not sure the statement is clear enough.  I think there are two points that need to clarify.  The 1st one is that there is a default value of  jdk.tls.maxCertificateChainLength, which is 10.  The 2nd one is that  jdk.tls.maxCertificateChainLength works for both client and server mode.  jdk.tls.server.maxInboundCertificateChainLength works on server mode, and it does not work for client mode, and therefore it cannot override client mode behavior for jdk.tls.maxCertificateChainLength.  The same for jdk.tls.client.maxInboundCertificateChainLength.

To be clear, the release note or the comment might be placed in different code block like:

```
/* If jdk.tls.server.maxInboundCertificateChainLength is set, it will override jdk.tls.maxCertificateChainLength behavior for server side.
 */

Integer inboundClientLen = ...

/* If jdk.tls.client.maxInboundCertificateChainLength is set, it will override jdk.tls.maxCertificateChainLength behavior for client side.
 */
 Integer inboundServerLen = GetIntegerAction.privilegedGetProperty(

```

>          * If neither jdk.tls.server.maxInboundCertificateChainLength nor
>          * jdk.tls.client.maxInboundCertificateChainLength is set, the behavior
>          * depends on the setting of jdk.tls.maxCertificateChainLength. If
>          * jdk.tls.maxCertificateChainLength is set, it falls back to that
>          * value; otherwise, it defaults to 8 for
>          * jdk.tls.server.maxInboundCertificateChainLength

Previously, the jdk.tls.maxCertificateChainLength default value is 10.  Now it is changed to 8.  This is a behavior change, please document it in release note, or just use 10 in the implementation.


>          * and 10 for jdk.tls.client.maxInboundCertificateChainLength.
>          * Usesrs can independently set either
>          * jdk.tls.server.maxInboundCertificateChainLength or
>          * jdk.tls.client.maxInboundCertificateChainLength.
>          */ 
> ```

[the remark this answers is not in the document]
Comment (seanjmullan) [reply src/java.base/share/classes/sun/security/ssl/SSLConfiguration.java]: I don't see a behavior change that conflicts with the CSR. I think it is a wording issue, let me suggest some improvements in another comment. There is no longer a default value for `jdk.tls.client.maxInboundCertificateChainLength`. Where is it set to 8 in the code?

[the remark this answers is not in the document]
Comment (seanjmullan) [reply src/java.base/share/classes/sun/security/ssl/SSLConfiguration.java]: I think the wording of the comment is somewhat confusing because it is trying to explain the behavior of both properties together and the words "either" and "neither" may be hard to parse. I recommend separate comment blocks for each property. Here is a suggestion for the server side setting:

```
/* 
 * maxInboundClientCertChainLen is the maximum length of a client certificate
 * chain accepted by a server. It is determined as follows:
 *  - If the jdk.tls.server.maxInboundCertificateChainLength system property
 *    is set and its value >= 0, it uses that value.
 *  - Otherwise, if the jdk.tls.maxCertificateChainLength system property is
 *    set and its value >= 0, it uses that value.
 *  - Otherwise it is set to a default value of 8.
 */
```

The client side setting would be similar.

[the remark this answers is not in the document]
Comment (haimaychao) [reply src/java.base/share/classes/sun/security/ssl/SSLConfiguration.java]: Yes, I can place the comments in the code blocks for the server-side setting and client-side setting, respectively.
@XueleiFan Any feedback before I'm making this comment change?
I will also update the release note accordingly. Thanks!

[the remark this answers is not in the document]
Comment (XueleiFan) [reply src/java.base/share/classes/sun/security/ssl/SSLConfiguration.java]: I'm not sure if there is a clear reason to change the default value from 10 to 8.  I'm fine if you want to keep to use value 10 for less compatibility issues. Otherwise, I have no more comment.  Thanks!

> Yes, I can place the comments in the code blocks for the server-side setting and client-side setting, respectively. @XueleiFan Any feedback before I'm making this comment change? I will also update the release note accordingly. Thanks!

[the remark this answers is not in the document]
Comment (seanjmullan) [reply src/java.base/share/classes/sun/security/ssl/SSLConfiguration.java]: The choice of 8 for the client is mostly based on different processing requirements and use cases for TLS client vs server certificate chains. If we see evidence that 8 is too low, we can always consider adjusting it.

[the remark this answers is not in the document]
Comment (XueleiFan) [reply src/java.base/share/classes/sun/security/ssl/SSLConfiguration.java]: I'm not sure if the number 8 or 10 really make a good difference in practice.  No matter 8 or 10,  if customers need lower value, they can always consider adjusting it.  My concern is mainly about compatibility issues.  If you want to keep the behavior changes, as there is potential compatibility issue, please feel free to describe the behavior change in release note and CSR if you would like.

[the remark this answers is not in the document]
Comment (seanjmullan) [reply src/java.base/share/classes/sun/security/ssl/SSLConfiguration.java]: Good point - the CSR and RN could have been a bit more specific about the compatibility effect of changing the default from 10 to 8, so we will update that. Note that the CertPathBuilder default max path length [is 5 non-self-issued intermediate CA certificates](https://docs.oracle.com/en/java/javase/21/docs/api/java.base/java/security/cert/PKIXBuilderParameters.html#setMaxPathLength(int)), so even a lower value of 8 should have low risk as rigid TLS cert chains greater than 6 certs (where the 6th cert is the end entity cert which is not affected by the CertPathBuilder limit) on the wire will already be rejected. The additional certs permitted in the wire format is more for including a few more additional certs that might help build a valid path when there is more than one possible chain that a server might accept, which sometimes happens.

[the remark this answers is not in the document]
Comment (XueleiFan) [reply src/java.base/share/classes/sun/security/ssl/SSLConfiguration.java]: Good point about the setMaxPathLength() limitation.  It looks like there might be an issue when the interactivities of setMaxPathLength(and its default value) and the properties defined here are not considered.  For example, what if the property is set to 8, while the setMaxPathLength is of value 5?  For the "PKIX" trust manager, per your description, if the property is set to 8, but 5 is the limit actually.  It looks like a weird behavior to me.  If I remember correctly, the "SunX509" trust manager does not use PKIXBuilderParameters, while the "PKIX" trust manager does.  It might be not the behavior we'd like to have that property 8 work for "SunX509" but not for "PKIX" trust manager.

Anyway, it might be better to look into the interactive behaviors among the properties and setMaxPathLength/default value.

> Good point - the CSR and RN could have been a bit more specific about the compatibility effect of changing the default from 10 to 8, so we will update that. Note that the CertPathBuilder default max path length [is 5 non-self-issued intermediate CA certificates](https://docs.oracle.com/en/java/javase/21/docs/api/java.base/java/security/cert/PKIXBuilderParameters.html#setMaxPathLength(int)), so even a lower value of 8 should have low risk as rigid TLS cert chains greater than 6 certs (where the 6th cert is the end entity cert which is not affected by the CertPathBuilder limit) on the wire will already be rejected. The additional certs permitted in the wire format is more for including a few more additional certs that might help build a valid path when there is more than one possible chain that a server or client might accept, which sometimes happens.

[the remark this answers is not in the document]
Comment (seanjmullan) [reply src/java.base/share/classes/sun/security/ssl/SSLConfiguration.java]: PKIX has been the default SunJSSE TrustManager since JDK 5 and AFAIK we have not had any reports of compatibility issues. This leads me to believe that valid TLS chains longer than 6 certificates are extremely rare.
